spec/src/main/asciidoc/variable-evaluation.asciidoc

```diff
@@ -0,0 +1,123 @@
+//
+// Copyright (c) 2016-2019 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// You may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+[[variable-evaluation]]
+== Variable Evaluation
+
+The value of a configuration property can contains a variable corresponding to another configuration property.
+This variable is _evaluated_ to be replaced by its own value when the original property is returned.
+
+For example, let's define a `server.url` property:
+
+[source]
+----
+server.url=http://${server.host}:{server.port}/endpoint
+----
+
+When the Config API returns the value of `server.url`, it evaluates any variable (such as `server.host` and `server.port`) to replace
+them with their value.
+
+If these properties are also defined:
+
+[source]
+----
+server.host=example.org
+server.port=8080
+----
+
+The `server.url` value returned by the Config API is `http://example.org:8080/endpoint`.
+
+[NOTE]
+====
+Backwards Compatibility
+
+Variable evaluation is not backwards compatible.
+Previous version of MicroProfile Config would not evalute the variables and would return the value `http://${server.host}:{server.port}/endpoint`.
+
+Implementation of MicroProfile Config MUST provide a way to disable variable evaluation to provide backwards compatibility.
+
+The property `mp.config.evaluateVariables` can be configured as a boolean in one of the 3 default `ConfigSource` (environment variables, system properties,
```

Comment:
-1 on specifying the location of the property. It should be allowed to be specified anywhere as a config, as long as it is available on application starting

Reply:
If I understand correctly, the property is supposed to apply on a per-source basis only, so that some sources can have expressions and some not.

Reply:
@emilyjiang we can not specify this property that is using to configure the `Config` itself on any `ConfigSource` or we end up with a chicken-and-egg situation.

Let's say we have a 3rd-party Zookeeper ConfigSource that calls `ConfigProvider.getConfig()` to get its own configuration (to be able to connect to ZK).

If we look for any `ConfigSource` to get this `mp.config.evaluateVariables`, we will call ZooKeep ConfigSource that will call `ConfigProvider.getConfig()` to get its configuration (to be able to fetch its properties). In turn this will call the method responsible to get the `mp.config.evaluateVariables` property, etc.

Restricting to the 3 default ConfigSource avoid this situation. This property is a special case: it it used to configure the `ConfigBuilder` itself, not the application. It seems ok to me to limit its use.